### PR TITLE
[Bug 1633754] Anomaly detection pipeline

### PR DESCRIPTION
**Pull request by scholtzan**
See https://bugzilla.mozilla.org/show_bug.cgi?id=1633754

Depends on https://github.com/mozilla/bigquery-etl/pull/949

I copied the code from https://github.com/mozilla/dscontrib/tree/master/src/dscontrib/jmccrosky/anomdtct and made some minor modifications to glue all the parts together. I was trying to find a way to cache and update the model, however it seems that this feature hasn't been implemented yet in `prophet` (see https://github.com/facebook/prophet/issues/46).

So instead, I added queries to bigquery-etl in https://github.com/mozilla/bigquery-etl/pull/949 which will incrementally run the queries from https://github.com/mozilla/dscontrib/blob/master/src/dscontrib/jmccrosky/anomdtct/data.py every day and store the results in BigQuery. The pipeline here will then query all the data from those tables and use it to calculate the models. The tables with the aggregated data are _a lot_ smaller than `clients_daily`, so that will not only increase speed but also save a lot of money.

I still need to run some tests to see if it works

### CODE REVIEW

File: anomdtct/entrypoint
Comment (scholtzan) [inline anomdtct/entrypoint]: This key should be set as configuration variable in Airflow

**Review by jmccrosky — COMMENTED** [review]
This is probably fine, but if you want, you can save the model from prophet (https://github.com/facebook/prophet/issues/725).  The other link referenced is for the case where we want to modify the model after saving and reloading it, which we don't need.

In theory, we can fit the model just once and each day we only need to make a prediction for that day using the model, load the one day of data, and add the one row (or set of rows, really) to the table.

What you have now may be cheap enough that it's not worth the trouble to further optimize though.  Your call :)

**Comment by jklukas** [comment]
> This is probably fine, but if you want, you can save the model from prophet ([facebook/prophet#725](https://github.com/facebook/prophet/issues/725)). 

That's probably feasible, though I think we should only look into this route if there's a clear performance/cost argument for it. Pickling as a serialization format has security implications; my understanding of this is that essentially a `pkl` file can contain arbitrary code that gets executed when you unpickle it. This is probably not a major problem as long as we have write access restricted to the location where we store the pickle. It might be best to store the bytes in a BQ table under `telemetry_derived` in which case only data ops and data eng would be able to modify the content.

File: anomdtct/entrypoint
Comment (jklukas) [reply anomdtct/entrypoint]: Do we know if command-line arguments show up in logs in Airflow?

Comment (jklukas) [reply anomdtct/entrypoint]: Even better, we should avoid sending a key at all. I assume we can set permissions on the spreadsheet to allow edits from the Airflow service account, so we can use the same creds as the container is using to access BQ.

**Review by jklukas — APPROVED** [review]
r+wc

Generally, this looks very close in structure to the existing simple_prohet lib, so I'm fine with merging what's here as long as we circle back to making sure this is hitting the new derived tables, isn't passing unnecessary creds, etc.

File: anomdtct/anomdtct/data.py
Comment (scholtzan) [inline anomdtct/anomdtct/data.py:49]: Currently blocked by https://github.com/mozilla/bigquery-etl/pull/952

File: anomdtct/anomdtct/data.py
Comment (scholtzan) [inline anomdtct/anomdtct/data.py:211]: @jmccrosky What happens if the model tries to perform forecasting for a region that it hasn't seen any training data for? Is it just discarded? This might be relevant in the `prepare_data` function below.

**Comment by scholtzan** [comment]
I updated the pipeline to cache generated models in BigQuery and reuse them for prediction.

File: anomdtct/anomdtct/data.py
Comment (jmccrosky) [inline anomdtct/anomdtct/data.py]: I guess we can turn this into a simple (not nested) query for cleanliness?

File: anomdtct/anomdtct/data.py
Comment (jmccrosky) [reply anomdtct/anomdtct/data.py:211]: The forecast function takes a dictionary as its data argument and expects training data to exist for every key in that dictionary.  If there are insufficient rows in the training data, it will skip that key though.  Looks like you've changed things a lot though :)  I think you're probably handling it fine.

File: anomdtct/anomdtct/data.py
Comment (jmccrosky) [inline anomdtct/anomdtct/data.py:225]: My code that enforces that we need at least 5000 dau per geoXday seems to be gone.  Is that somewhere else?  It's a requirement from data stewardship.

Also, droping geos for which we don't have at least 600 days of training data left.

**Review by jmccrosky — COMMENTED** [review]
Looking nice :)

File: anomdtct/anomdtct/data.py
Comment (scholtzan) [reply anomdtct/anomdtct/data.py:225]: Right, I added the line back that at least 5000dau are required.

The 600 days of training data are handled in `prepare_training_data`. If there is less data then we don't compute a model for that geo. So we won't do predictions for geos with less than 600 days of data since no model actually exist for those.

**Comment by scholtzan** [comment]
https://github.com/mozilla/bigquery-etl/pull/952 adds `client_first_seen`, so we are no longer blocking on this.

File: anomdtct/anomdtct/data.py
Comment (jmccrosky) [reply anomdtct/anomdtct/data.py:225]: Cool :)

**Review by jmccrosky — APPROVED** [review]
So cool!  Thanks for doing this.  As far as I can tell, it's all good.

File: anomdtct/anomdtct/pipeline.py
Comment (jklukas) [inline anomdtct/anomdtct/pipeline.py]: How many models are we fitting in this loop, and how large are the pickled models? We may want to build up a list of these pickled models and save them all in one load job. We could have load jobs fail if we try to run many of them in quick succession.

File: anomdtct/anomdtct/pipeline.py
Comment (jklukas) [inline anomdtct/anomdtct/pipeline.py]: The production pipeline should be writing into tables in the `telemetry_derived` dataset and loading cached models from `telemetry_derived`. We can do this now since this is going to be run via Airflow and have the necessary creds.

I would like to see the production defaults written in here, and users running this locally would be expected to provision their own tables in the DS project or in `analysis` and configure the job accordingly.

File: anomdtct/fit_models
Comment (jklukas) [inline anomdtct/fit_models]: This docstring doesn't make sense for the content here.

**Review by jklukas — APPROVED** [review]
r+wc

**Comment by scholtzan** [comment]
I got the spreadsheet export working, however the first time I ran it I had to manually activate the API for the account and share the spreadsheet with the account:

```
googleapiclient.errors.HttpError: <HttpError 403 when requesting https://sheets.googleapis.com/v4/spreadsheets/.../values/Sheet1?valueInputOption=RAW&alt=json returned "Google Sheets API has not been used in project <project_id> before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/sheets.googleapis.com/overview?project=<project_id> then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.". 
```

This is something which also needs to be done for the Airflow account when this stuff gets set up there.

**Comment by scholtzan** [comment]
Models are now cached in `moz-fx-data-shared-prod.telemetry_derived.deviations_model_cache_v1`. I also tested prediction for one day of data. It all seems to work.

**Pull request merged**
